src/registrar_agent.rs

```diff
@@ -0,0 +1,95 @@
+use crate::error::Error;
+use reqwest::header::*;
+use serde::{Deserialize, Serialize};
+use serde_json::Number;
+
+fn serialize_as_base64<S>(
+    bytes: &[u8],
+    serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+{
+    serializer.serialize_str(&base64::encode(bytes))
+}
+
+fn deserialize_as_base64<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    String::deserialize(deserializer).and_then(|string| {
+        base64::decode(&string).map_err(serde::de::Error::custom)
+    })
+}
+
+pub fn deserialize_maybe_base64<'de, D>(
+    deserializer: D,
+) -> Result<Option<Vec<u8>>, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    Option::<WrappedBase64Encoded>::deserialize(deserializer)
+        .map(|wrapped| wrapped.map(|wrapped| wrapped.0))
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Register<'a> {
+    ek: &'a [u8],
+    #[serde(serialize_with = "serialize_as_base64")]
+    ekcert: &'a [u8],
+    #[serde(serialize_with = "serialize_as_base64")]
+    ek_tpm: &'a [u8],
+    aik: &'a [u8],
+    aik_name: &'a [u8],
+    tpm_version: u8,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct RegisterResponseResults {
+    #[serde(deserialize_with = "deserialize_maybe_base64")]
+    blob: Option<Vec<u8>>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct RegisterResponse {
+    code: Number,
+    status: String,
+    results: RegisterResponseResults,
+}
+
+#[derive(Debug, Deserialize)]
+struct WrappedBase64Encoded(
+    #[serde(deserialize_with = "deserialize_as_base64")] Vec<u8>,
+);
+
+#[allow(clippy::too_many_arguments)]
+pub(crate) async fn do_register_agent(
+    registrar_ip: &str,
+    registrar_port: &str,
+    agent_uuid: &str,
+    ek: &[u8],
+    ekcert: &[u8],
+    aik: &[u8],
+    ek_tpm: &[u8],
+    aik_name: &[u8],
+) -> Result<RegisterResponse, Error> {
+    let data = Register {
+        ek,
+        ekcert,
+        ek_tpm,
+        aik,
+        aik_name,
+        tpm_version: 2,
+    };
+    reqwest::Client::new()
+        .post(&format!(
+            "http://{}:{}/agents/{}",
+            registrar_ip, registrar_port, agent_uuid
+        ))
+        .json(&data)
+        .send()
+        .await?
+        .json()
```

Comment:
This will fail if the server returned an error, because then `code` is non-200 and `status` is not `success`, and it won't have a `results` block.

Reply:
I moved the status code check to be before this `.json()` is called and just used the built-in `reqwest::StatusCode` method.